sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py

```diff
@@ -765,6 +767,27 @@ def _group_stages_by_key(stages, get_stage_key):
   return (grouped_stages, stages_with_none_key)
 
 
+def _group_stages_with_limit(stages, get_limit):
+  # type: (Iterable[Stage], Callable[[[str], int]]) -> Iterable[Iterable[Stage]]
+  stages_with_limit = [(stage, get_limit(stage.name)) for stage in stages]
+  group = []
+  group_limit = 0
+  for stage, limit in sorted(stages_with_limit, key=operator.itemgetter(1)):
+    if limit < 1:
+      raise Exception(
+          'expected get_limit to return an integer >= 1, '
+          'instead got: %d for stage: %s' % (limit, stage))
+    if not group:
+      group_limit = limit
+    assert len(group) < group_limit
+    group.append(stage)
```

Comment:
This will put stages into groups that are too big, e.g. if we have

`[(big, 10), (s1, 2), (s2, 2)]`

they will all end up in the same group. Instead, it seems you should append if len(group) < limit, else make a new group.

Reply:
I ran this manually and the output was `[['s1', 's2'], ['big']]`. I added a similar test case for testing a mix of limits.

Reply:
My mistake, I was thinking the iteration was from largest to smallest.